dataactvalidator/validation_handlers/validationManager.py

```diff
@@ -416,17 +430,73 @@ def load_file_data(self, sess, bucket_name, region_name):
 
         return file_row_count
 
-    def process_data_chunk(self, sess, chunk_df):
+    def parallel_data_loading(self, reader_obj):
+        with Manager() as server_manager:
+            # These variables will need to be shared among the processes and used later overall
+            shared_data = server_manager.dict()
+            shared_data['total_rows'] = self.total_rows
+            shared_data['has_data'] = self.has_data
+            shared_data['error_rows'] = self.error_rows
+            shared_data['error_list'] = self.error_list
+            shared_data['header_dict'] = self.reader.header_dict
+            shared_data['flex_fields'] = self.reader.flex_fields
+            # setting reader to none as multiprocess can't pickle it, it'll get reset
+            temp_reader = self.reader
+            self.reader = None
+
+            m_lock = server_manager.Lock()
+            pool = Pool(MULTIPROCESSING_POOLS)
+            for chunk_df in reader_obj:
+                pool.apply_async(func=self.process_data_chunk, args=(chunk_df, shared_data, m_lock))
+            pool.close()
+            pool.join()
+
+            # Resetting these out here as they are used later in the process
+            self.total_rows = shared_data['total_rows']
+            self.has_data = shared_data['has_data']
+            self.error_rows = shared_data['error_rows']
+            self.error_list = shared_data['error_list']
+            self.reader = temp_reader
+
+    def iterative_data_loading(self, reader_obj):
+        shared_data = {
+            'total_rows': self.total_rows,
+            'has_data': self.has_data,
+            'error_rows': self.error_rows,
+            'error_list': self.error_list,
+            'header_dict': self.reader.header_dict,
+            'flex_fields': self.reader.flex_fields
+        }
+        for chunk_df in reader_obj:
+            self.process_data_chunk(chunk_df, shared_data)
+
+        # Resetting these out here as they are used later in the process
+        self.total_rows = shared_data['total_rows']
+        self.has_data = shared_data['has_data']
+        self.error_rows = shared_data['error_rows']
+        self.error_list = shared_data['error_list']
+
+    def process_data_chunk(self, chunk_df, shared_data, m_lock=None):
         """ Loads in a chunk of the file and performs initial validations
 
             Args:
-                sess: the database connection
                 chunk_df: the chunk of the file to process as a dataframe
+                shared_data: dictionary of shared data among the chunks
+                m_lock: manager lock if provided to ensure processes don't override each other
         """
+        if m_lock:
+            # make a new connection per process
+            conn = db_connection()
```

Comment:
Since the database transaction is an attribute of a connection, this should cause each chunk being processed in an iteration to run in a separate transaction (and separate SQLAlchemy `Session` object, which is not what you want. 
- When in parallel, failure of one chunk won't rollback data from other chunks that have succeeded

I spend a good deal of time ensuring that db sessions started in a parent process could continue to be used in child worker processes the the sqs dispatcher and child workers. I think you should be able to do that here, but I'm not sure if transactions can cross process boundaries and still work. This may make for tricky overall coordination - and perhaps needs some research on "coordinating sqlalchemy transactions over multiprocessing pool"

Reply:
You could test this out by running a large file in parallel and killing it halfway ... then see if all data was rolled back or not.

Reply:
Yeah I ran into many deadlocks using the same db session/connection objects and settled on this way after most of the research concluding something along these lines (https://docs-sqlalchemy.readthedocs.io/ko/latest/faq/connections.html#how-do-i-use-engines-connections-sessions-with-python-multiprocessing-or-os-fork):
 * The db socket gets confused when split/shared into multiple processes.
 * So either remove the poolling from the db engine or sever all connections before splitting and setting them up in the processes.
I _am_ curious how you currently have it setup with the dispatcher though so I'll look into that.

Reply:
Updated so that it uses temp tables in case something goes wrong with the data loading.